File: src/test/scala/com/kotori316/scala_lib/util/ModClassDataTest.scala
package com.kotori316.scala_lib.util

import cats.data.Validated
import cats.data.Validated.Invalid
import com.kotori316.scala_lib.ModClassData
import com.kotori316.scala_lib.ModClassData.Duplicated
import org.junit.jupiter.api.Assertions._
import org.junit.jupiter.api.Test

//noinspection SpellCheckingInspection
private[util] class ModClassDataTest {
  private[this] final val forge = Data("net.minecraftforge.common.ForgeMod", "forge")
  private[this] final val mc = Data(classOf[String].getName, "minecraft")
  private[this] final val quarry = Data("com.yogpc.qp.QuarryPlus", "quarryplus")
  private[this] final val tank = Data("com.kotori316.fluidtank.FluidTank ", "fluidtank")
  private[this] final val jei = Data("mezz.jei.JustEnoughItems   ", "jei")
  val normalCase = List(forge, mc, quarry, tank, jei)

  @Test
  private[util] def findInstanceNormal(): Unit = {
    val value = ModClassData.findInstance(normalCase)
    assertTrue(value.isValid, s"Normal is valid, $value")
  }

  @Test
  def findInstance2(): Unit = {
    val data = normalCase :+ quarry.copy(className = quarry.className + "$")
    val instances = ModClassData.findInstance(data)
    assertAll(
      () => assertTrue(instances.isValid, s"Including Object, $data"),
      () => assertTrue(instances.exists(_.find(_.modID == "quarryplus").exists(_.isScalaObj)))
    )
  }

  @Test
  def findInstance3(): Unit = {
    val data = normalCase :+ quarry.copy(className = "String")
    val instances = ModClassData.findInstance(data)
    assertAll(
      () => assertTrue(instances.isInvalid, s"Duplicated, $data"),
      () => instances match {
        case Invalid(e) => assertTrue(e.contains(Duplicated("quarryplus")), s"error in quarry, $e")
        case v => fail(s"What happened to $v")
      }
    )
  }

  @Test
  def findInstance4(): Unit = {
    val data = normalCase :+
      quarry.copy(className = quarry.className + "$") :+
      quarry.copy(className = "java.lang.Object" + "$")
    val instances = ModClassData.findInstance(data)
    assertAll(
      () => assertTrue(instances.isInvalid, s"Including 2 Objects, $data"),
      () => instances match {
        case Invalid(e) => assertTrue(e.contains(Duplicated("quarryplus")), s"error in quarry, $e")
        case v => fail(s"What happened to $v")
      }
    )
  }

  @Test
  def dummy(): Unit = {
    assertTrue(true)
    assertEquals(1, 3 - 2)
  }

  @Test
  def easyGoodCase1(): Unit = {
    val oneClass = List(
      Data("com.kotori316.Mod", "my_mod")
    )
    val oneInstance = ModClassData.findInstance(oneClass)
    assertEquals(Validated.valid(List(Data("com.kotori316.Mod", "my_mod"))), oneInstance)
  }

  @Test
  def easyGoodCase1_1(): Unit = {
    val fiveClasses = List(
      Data("com.kotori316.Mod1", "my_mod1"),
      Data("com.kotori316.Mod2", "my_mod2"),
      Data("com.kotori316.Mod3", "my_mod3"),
      Data("com.kotori316.Mod4", "my_mod4"),
      Data("com.kotori316.Mod5", "my_mod5"),
    )
    val fiveInstances = ModClassData.findInstance(fiveClasses).map(_.sortBy(_.modID))
    assertEquals(Validated.valid(fiveClasses), fiveInstances)
  }

  @Test
  def easyGoodCase2(): Unit = {
    val classAndObject = List(
      Data("com.kotori316.Mod", "my_mod"), // Just a class
      Data("com.kotori316.Mod$", "my_mod") // An Object
    )
    val validatedObject = ModClassData.findInstance(classAndObject)
    assertEquals(Validated.valid(List(Data("com.kotori316.Mod$", "my_mod"))), validatedObject)
  }

  @Test
  def easyErrorCase1(): Unit = {
    val mods = List(
      Data("com.kotori316.Mod1", "my_mod1"),
      Data("com.kotori316.Mod_Extended", "my_mod1"),
    )
    val validiation = ModClassData.findInstance(mods)
    assertAll(
      () => assertTrue(validiation.isInvalid),
      () => assertTrue(validiation.swap.exists(_.contains(Duplicated("my_mod1"))))
    )
  }

  case class Data(override val className: String, override val modID: String) extends ModClassData

}

File: neoforge/src/test/java/com/kotori316/scala_lib/ModClassDataTest.java
package com.kotori316.scala_lib;

import java.util.List;
import java.util.Set;

import org.junit.jupiter.api.Nested;
import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.assertDoesNotThrow;
import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.junit.jupiter.api.Assertions.assertTrue;

class ModClassDataTest {
    @Test
    void isScalaObj() {
        assertTrue(new ModClassDataImpl("com.kotori316.test.Mod$", "test1").isScalaObj());
    }

    @Test
    void isClass() {
        assertFalse(new ModClassDataImpl("com.kotori316.test.Mod", "test1").isScalaObj());
    }

    @Nested
    class FindInstanceTest {
        @Test
        void justOneClass() {
            var target = new ModClassDataImpl("com.kotori316.test.Mod", "test1");
            var result = assertDoesNotThrow(() -> ModClassData.findInstance(List.of(target)));
            assertEquals(List.of(target), result);
        }

        @Test
        void justOneObj() {
            var target = new ModClassDataImpl("com.kotori316.test.Mod$", "test1");
            var result = assertDoesNotThrow(() -> ModClassData.findInstance(List.of(target)));
            assertEquals(List.of(target), result);
        }

        @Test
        void twoModId() {
            var targets = Set.of(new ModClassDataImpl("com.kotori316.test.Mod1", "test1"),
                new ModClassDataImpl("com.kotori316.test.Mod2", "test2"));
            var result = Set.copyOf(assertDoesNotThrow(() -> ModClassData.findInstance(targets)));
            assertEquals(targets, result);
        }

        @Test
        void threeModIdWithObject() {
            var targets = Set.of(
                new ModClassDataImpl("com.kotori316.test.Mod1$", "test1"),
                new ModClassDataImpl("com.kotori316.test.Mod1", "test1"),
                new ModClassDataImpl("com.kotori316.test.Mod2", "test2"),
                new ModClassDataImpl("com.kotori316.test.Mod3$", "test3"),
                new ModClassDataImpl("com.kotori316.test.Mod3", "test3"),
                new ModClassDataImpl("com.kotori316.test.Mod3_1", "test3")
            );
            var result = Set.copyOf(assertDoesNotThrow(() -> ModClassData.findInstance(targets)));
            assertEquals(Set.of(
                new ModClassDataImpl("com.kotori316.test.Mod1$", "test1"),
                new ModClassDataImpl("com.kotori316.test.Mod2", "test2"),
                new ModClassDataImpl("com.kotori316.test.Mod3$", "test3")
            ), result);
        }

        @Test
        void classAndObj() {
            var targets = Set.of(
                new ModClassDataImpl("com.kotori316.test.Mod1$", "test1"),
                new ModClassDataImpl("com.kotori316.test.Mod1", "test1")
            );
            var result = Set.copyOf(assertDoesNotThrow(() -> ModClassData.findInstance(targets)));
            assertEquals(Set.of(new ModClassDataImpl("com.kotori316.test.Mod1$", "test1")), result);
        }

        @Test
        void classAndObj2() {
            var targets = Set.of(
                new ModClassDataImpl("com.kotori316.test.Mod1$", "test1"),
                new ModClassDataImpl("com.kotori316.test.Mod1", "test1"),
                new ModClassDataImpl("com.kotori316.test.Mod1_other", "test1")
            );
            var result = Set.copyOf(assertDoesNotThrow(() -> ModClassData.findInstance(targets)));
            assertEquals(Set.of(new ModClassDataImpl("com.kotori316.test.Mod1$", "test1")), result);
        }

        @Test
        void duplication1() {
            var targets = List.of(
                new ModClassDataImpl("com.kotori316.test.Mod1", "test1"),
                new ModClassDataImpl("com.kotori316.test.Mod1", "test1")
            );
            assertThrows(RuntimeException.class, () -> ModClassData.findInstance(targets));
        }

        @Test
        void duplication2() {
            var targets = Set.of(
                new ModClassDataImpl("com.kotori316.test.Mod1", "test1"),
                new ModClassDataImpl("com.kotori316.test.Mod2", "test1")
            );
            assertThrows(RuntimeException.class, () -> ModClassData.findInstance(targets));
        }

        @Test
        void duplication3() {
            var targets = Set.of(
                new ModClassDataImpl("com.kotori316.test.Mod1", "test1"),
                new ModClassDataImpl("com.kotori316.test.Mod2", "test1"),
                new ModClassDataImpl("com.kotori316.test.Mod3", "test1")
            );
            assertThrows(RuntimeException.class, () -> ModClassData.findInstance(targets));
        }

        @Test
        void duplication4() {
            var targets = Set.of(
                new ModClassDataImpl("com.kotori316.test.Mod1", "test1"),
                new ModClassDataImpl("com.kotori316.test.Mod2", "test2"),
                new ModClassDataImpl("com.kotori316.test.Mod3", "test1")
            );
            assertThrows(RuntimeException.class, () -> ModClassData.findInstance(targets));
        }
    }

    record ModClassDataImpl(String className, String modID) implements ModClassData {
    }
}